src/main/java/org/gradle/profiler/maven/MavenScenarioInvoker.java

```diff
@@ -1,25 +1,53 @@
 package org.gradle.profiler.maven;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import org.gradle.profiler.BuildToolCommandLineInvoker;
-import org.gradle.profiler.GradleBuildConfiguration;
+import org.gradle.profiler.InstrumentingProfiler;
+import org.gradle.profiler.InstrumentingProfiler.SnapshotCapturingProfilerController;
 import org.gradle.profiler.InvocationSettings;
-import org.gradle.profiler.Logging;
+import org.gradle.profiler.ScenarioSettings;
 import org.gradle.profiler.result.BuildInvocationResult;
 
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Consumer;
 
 public class MavenScenarioInvoker extends BuildToolCommandLineInvoker<MavenScenarioDefinition, BuildInvocationResult> {
     @Override
-    public void run(MavenScenarioDefinition scenario, InvocationSettings settings, Consumer<BuildInvocationResult> resultConsumer) {
+    public void run(MavenScenarioDefinition scenario, InvocationSettings settings, Consumer<BuildInvocationResult> resultConsumer) throws IOException, InterruptedException {
         List<String> commandLine = new ArrayList<>();
         commandLine.add(scenario.getExecutablePath());
         commandLine.addAll(scenario.getTargets());
         scenario.getSystemProperties().forEach((key, value) ->
             commandLine.add(String.format("-D%s=%s", key, value)));
 
-        doRun(scenario, settings, resultConsumer, commandLine);
+        Map<String, String> profileEnvironment;
+        SnapshotCapturingProfilerController controller;
+        // TODO This only works with Async profiler, since the only thing we call from the controller is stopSession()
+        //      Capture this in the type hierarchy somehow
+        if (settings.getProfiler() instanceof InstrumentingProfiler) {
```

Comment:
🤔 Should we throw exception if profiler is not supported?

Maybe a profiler would have something like:
`Profiler.supports(BuildTool)`

Reply:
I tried that, but it got a bit ugly. I think the answer is to remodel the relationship between profilers and build tools on a bit deeper level, but it's a more impactful change than I think will fit in this PR.

Reply:
Let's continue on https://github.com/gradle/gradle-profiler/pull/576